### PR TITLE
RFC: sRGB filter converts images to sRGB using embedded profiles

### PR DESCRIPTION
**Pull request by kkopachev**
Hi! 
Please don't merge, as this is mostly proof of concept to get your comments. I'll add config option and tests, if main code is good enough.

New srgb filter converts images with embedded profiles to sRGB, embedding new tiny sRGB profile (could be stripped later with strip_icc filter)
That should solve issues with CMYK images as well as allow to replace embedded image profiles with tiny sRGB one. 

related issues: https://github.com/thumbor/thumbor/issues/586 https://github.com/thumbor/thumbor/issues/740

Noot sure if it's a good way implementing this, though. Currently filter relies on internal structure of Pil engine, because there is no another way to get image from the engine. `get_image_data` method does not provide access to embedded profile and `image_data_as_rgb` converts image to RGB without proper color management.

Maybe that should all be implemented inside engine and exposed to filter similar as `strip_icc`, but I'm afraid that could create BC break.

Thoughts?

### CODE REVIEW

**Comment by vasekboch** [comment]
Exactly. I was trying to write same filter. But stumbled on this line.  There is a implicit conversion here https://github.com/thumbor/thumbor/blob/master/thumbor/engines/pil.py#L186. I So inside filter  self.engine.image.mode returns RGB. So its impossible to apply the embedded profile. Its possible to convert it back to CMYK (I can detect it from ICC profile) but still some data are lost. This needs to be fixed in ```thumbor/engines/pil.py```. And I don't know Thumbor that well, to do so.
Basically the conversion without applying ICC profile resolved to those issues  #586 #740

File: thumbor/filters/srgb.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                       3, 22, 0, 0, 3, 51, 0, 0, 2, 164, 88, 89, 90, 32, 0, 0, 0, 0, 0, 0, 111, 162, 0, 0, 56, 245, 0,
+                       0, 3, 144, 88, 89, 90, 32, 0, 0, 0, 0, 0, 0, 98, 153, 0, 0, 183, 133, 0, 0, 24, 218, 88, 89, 90,
+                       32, 0, 0, 0, 0, 0, 0, 36, 160, 0, 0, 15, 132, 0, 0, 182, 207, 99, 117, 114, 118, 0, 0, 0, 0, 0,
+                       0, 0, 26, 0, 0, 0, 203, 1, 201, 3, 99, 5, 146, 8, 107, 11, 246, 16, 63, 21, 81, 27, 52, 33, 241,
+                       41, 144, 50, 24, 59, 146, 70, 5, 81, 119, 93, 237, 107, 112, 122, 5, 137, 177, 154, 124, 172,
+                       105, 191, 125, 211, 195, 233, 48, 255, 255])
+
+
+class Filter(BaseFilter):
+    phase = thumbor.filters.PHASE_AFTER_LOAD
```
Comment (kkopachev) [inline thumbor/filters/srgb.py]: Filter is applied right after load, before engine does it's work.

**Comment by kkopachev** [comment]
@vasekboch Not sure I understand why would you like to convert image back to CMYK?
That line you pointed out is not only place where forced RGB conversion happens. Filters that call `image_data_as_rgb` will force RGB conversion too.

Apart from hackiness, proposed filter is working.

**Comment by vasekboch** [comment]
Well the problem is, that when I get the image inside the plugin its already in RGB with implicit conversion that does not care about ICCs. For example this image does not work.

plane

I used this filter, installed it and tried my problematic image. But the result is this error.

```
2017-06-01 19:54:43 tornado.application:ERROR Future exception was never retrieved: Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/tornado/gen.py", line 1069, in run
    yielded = self.gen.send(value)
  File "/usr/local/lib/python2.7/site-packages/thumbor/handlers/__init__.py", line 207, in get_image
    self.filters_runner.apply_filters(thumbor.filters.PHASE_AFTER_LOAD, transform)
  File "/usr/local/lib/python2.7/site-packages/thumbor/filters/__init__.py", line 91, in apply_filters
    exec_one_filter()
  File "/usr/local/lib/python2.7/site-packages/thumbor/filters/__init__.py", line 90, in exec_one_filter
    f.run(exec_one_filter)
  File "/usr/local/lib/python2.7/site-packages/thumbor/filters/__init__.py", line 198, in run
    results.append(self.runnable_method(*self.params))
  File "/usr/local/lib/python2.7/site-packages/thumbor/filters/__init__.py", line 23, in wrapper
    return fn(self, *args2)
  File "/usr/local/lib/python2.7/site-packages/thumbor_srgb_filter/srgb.py", line 74, in srgb
    outputMode=output_mode)
  File "/usr/local/lib/python2.7/site-packages/PIL/ImageCms.py", line 355, in profileToProfile
    raise PyCMSError(v)
PyCMSError: cannot build transform
```

I was digging around and found out that, ```self.engine.image.mode``` returns RGB. That is the reason why the profileToProfile fails. Due to the implicit conversion.

I changed the lines to this:
```
transform = ImageCms.buildTransform(embedded_profile, srgb_profile, "CMYK", output_mode)
image = ImageCms.applyTransform(image, transform)
```
This returns obvious nonsense. But applyTransformation works withnout error. Tweaking code to 
```
image = image.convert("CMYK")
transform = ImageCms.buildTransform(embedded_profile, srgb_profile, "CMYK", output_mode)
image = ImageCms.applyTransform(image, transform)
```
Works for this example, but the colors are a little bit wrong due to CMYK -> RBG -> CMYK -> RBG conversion.

Running it like this works correctly as expected.
```
from PIL import Image
from PIL import ImageCms
from io import BytesIO

image = Image.open('plane.jpg')

print(image.mode)

embedded_profile = image.info.get('icc_profile')
embedded_profile = BytesIO(embedded_profile)

srgb_profile = ImageCms.createProfile("sRGB")
transform = ImageCms.buildTransform(embedded_profile, srgb_profile, image.mode, "RGB")
image = ImageCms.applyTransform(image, transform)

print(image.mode) # Output: RGB

image.save('plane_new.jpg')
```
 You see why am I asking about the implicit conversion to sRGB. For this image it does not work correctly due to the ICC and CMYK combo. :)

**Comment by kkopachev** [comment]
That's weird, because it works just fine on my setup. No errors and output in srgb. I specifically tested filter for CMYK image with embedded ICC profile. 
http://localhost:8001/unsafe/fit-in/1024x1024/filters:srgb()/https://cloud.githubusercontent.com/assets/1482932/26698213/5fa2705c-4715-11e7-9570-072611fad259.jpg

**Comment by vasekboch** [comment]
Ahhh OK, well that is really weird. 
Before I was running 6.3.0. I tried to upgrade to 6.3.2. And even your link throws the same error.
What could be wrong, that would cause this behaviour? 
I've just installed fresh thumbor through pip and created simple setup.py for the srgb.py file and installed it as package. All in default configuration with added FILTER.

**Comment by vasekboch** [comment]
@kkopachev I was digging around and located my issue to ```STORAGE="thumbor.storages.file_storage"``` If I use ```STORAGE="thumbor.storages.no_storage"``` option everything works as expected. But when I set storage to default eq. ```thumbor.storages.file_storage``` it fails on error that I was receiving before. With no_storage works as a charm. I was suing everything on default, so that's why it was failing. 

I would love to help, but my thumbor and python knowledge is very limited. But If I found anything new, I will let you know.

**Comment by kkopachev** [comment]
@vasekboch found it: https://github.com/thumbor/thumbor/pull/930

**Comment by vasekboch** [comment]
@kkopachev Cool, thx man. We will test it if has some side effects. :)

**Comment by masom** [comment]
@kkopachev I think this should probably be embedded in the pil engine and enabled with a config variable.

Knowing if the image is CMYK requires prior knownledge at upload time. It would be great if the pil engine could just detect the CMYK mode and automatically transform to sRGB when enabled.

Having it as a filter is also a good option but it could just call in the pil engine to apply the profile conversion?

**Comment by JOGUI22** [comment]
@masom yeah, i think just the same. Could be awesome if thumbor automatically change CMYK profiles to sRGB.

I think is quite an important issue, because when an user uploads an image like this one in CMYK:

cystart

Thumbor outputs this version, with changed colors:

cyend

**Comment by znerol** [comment]
I have thumbor deployed as part of an automated print-first workflow. I resolved exact the same issue with a custom filter as well: https://github.com/znerol/thumbor-icc

I also think that `image_data_as_rgb` should apply embedded color profiles (if PIL is compiled with color management support).

**Comment by kkopachev** [comment]
@znerol could you explain a little more what do you mean by 
> I also think that image_data_as_rgb should apply embedded color profiles

**Comment by znerol** [comment]
https://xkcd.com/1882/

xkcd

**Comment by heynemann** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- thumbor/filters/srgb.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/heynemann/thumbor/pullRequest?prid=4967917&bid=16414100)

**Comment by karptonite** [comment]
This definitely looks like something we could use--we are running into issues with a small number of user provided images with gigantic (700k) ICC profiles. I don't want to strip them willy-nilly, of course, but I also don't want a few giant images cause issues for our users. Is this on track to be added to Thumbor, or should I instead look at @znerol's custom filter?

**Review by lastthyme — APPROVED** [review]

**Comment by heynemann** [comment]
Closing this PR since we had no activity in two years in the interest of keeping our PR list leaner. If this is something we still want to pursue, please reopen and I'll look into it. Thanks!
